### PR TITLE
Implement function type descriptors and values (part 1)

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
At this stage, we will support function values without variance.
Related to #907 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

**Comment by heshanpadmasiri** [comment]
@jclark and @manuranga  there are two places I am not sure what the output should be
1. When printing a function value directly ([`15-function/print-v.bal`](https://github.com/ballerina-platform/nballerina/blob/6c22b37cbd6eec55c8faa1b8822f8c8d9742c861/compiler/testSuite/15-function/print-v.bal)
2. What should be the sExpr for function value [here](https://github.com/ballerina-platform/nballerina/blob/6c22b37cbd6eec55c8faa1b8822f8c8d9742c861/compiler/modules/bir.sexpr/serialize.bal#L155)

(If you have `F f = foo; // where foo is a function` Right hand side of the assign instruction will have a `bir:FunctionValOperand`)

Other than that rest of the implementation is ready for review.

**Comment by manuranga** [comment]
Please test indirect calling of `array:push` function.

File: compiler/modules/bir/bir.bal [hunk 1]
```diff
@@ -385,12 +389,17 @@ public type StringConstOperand readonly & record {|
     string value;
 |};
 
+public type FunctionConstValOperand readonly & record {|
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: FunctionConstOperand

File: compiler/modules/bir/bir.bal [hunk 2]
```diff
@@ -596,10 +605,27 @@ public type CallInsn readonly & record {|
...
 |};
 
+# Call a function using a function value.
+# This behaves similar to CallInsn.
+public type CallIndirectInsn readonly & record {|
+    *ResultInsnBase;
+    INSN_CALL_INDIRECT name = INSN_CALL_INDIRECT;
+    Register func;
+    FunctionRef funcRef;
+    Operand[] args;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: Operand[] operands;

Comment (jclark) [reply compiler/modules/bir/bir.bal]: `[Register, Operand...]` operands

File: compiler/modules/bir/bir.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: Fixed

File: compiler/modules/bir/bir.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: Fixed

File: compiler/modules/bir/bir.bal [hunk 3]
```diff
@@ -385,12 +388,17 @@ public type StringConstOperand readonly & record {|
...
+    FunctionRef value;
+|};
+
 public type IntOperand IntConstOperand|Register;
 public type FloatOperand FloatConstOperand|Register;
 public type DecimalOperand DecimalConstOperand|Register;
 public type BooleanOperand BooleanConstOperand|Register;
 public type StringOperand StringConstOperand|Register;
-public type FunctionOperand FunctionRef|Register;
+public type FunctionOperand FunctionRef|FunctionConstOperand|Register;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: Get rid of this.

Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: Fixed

File: compiler/modules/bir/bir.bal [hunk 4]
```diff
@@ -342,13 +343,15 @@ public type Insn
...
     |MappingConstructInsn|MappingGetInsn|MappingSetInsn
-    |StringConcatInsn|RetInsn|AbnormalRetInsn|CallInsn
+    |StringConcatInsn|RetInsn|AbnormalRetInsn|CallInsn|CallIndirectInsn
     |AssignInsn|TypeCastInsn|TypeTestInsn|TypeMergeInsn
     |BranchInsn|TypeBranchInsn|CondBranchInsn|CatchInsn|PanicInsn|ErrorConstructInsn;
 
 public type Operand ConstOperand|Register;
 
-public type ConstOperand  readonly & record {|
+public type ConstOperand SingleValueConstOperand|FunctionConstOperand;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal:352]: Make this more explicit.

**Review by jclark — APPROVED** [review]

**Pull request merged**